From 96338dfb6ab66938635c26844597f72464fccc0a Mon Sep 17 00:00:00 2001
From: Sidhanta Sahu <quic_sidhanta@quicinc.com>
Date: Mon, 29 Jan 2024 16:40:32 -0800
Subject: [PATCH] hostapd: Set sta wpa sm to WPA_DISASSOC on low ack

During low ack event, existing sta wpa_sm is not set to DISASSOC while
dissociating the sta.

Set the sta wpa sm to WPA_DISAASOC to avoid unexpected behavior due to
deauthentication reason as CLASS3_FRAME_FROM_NONASSOC_STA.

Patch-depdendency:none
Patch-work:none

Signed-off-by: Sidhanta Sahu <quic_sidhanta@quicinc.com>
---
 src/ap/sta_info.c | 7 +++++--
 1 file changed, 5 insertions(+), 2 deletions(-)

--- a/src/ap/sta_info.c
+++ b/src/ap/sta_info.c
@@ -1079,6 +1079,7 @@ void ap_sta_disassociate(struct hostapd_
 	}
 	ap_sta_set_authorized(hapd, sta, 0);
 	hostapd_set_sta_flags(hapd, sta);
+	wpa_auth_sm_event(sta->wpa_sm, WPA_DISASSOC);
 	wpa_printf(MSG_DEBUG, "%s: reschedule ap_handle_timer timeout "
 		   "for " MACSTR " (%d seconds - "
 		   "AP_MAX_INACTIVITY_AFTER_DISASSOC)",
@@ -1089,6 +1090,7 @@ void ap_sta_disassociate(struct hostapd_
 			       ap_handle_timer, hapd, sta);
 	accounting_sta_stop(hapd, sta);
 	ieee802_1x_free_station(hapd, sta);
+	sta->added_unassoc = 0;
 #ifdef CONFIG_IEEE80211BE
 	if (!hapd->conf->mld_ap ||
 	    hapd->mld_link_id == sta->mld_assoc_link_id) {
